rear_camera_py/src/line_angle_calculator.py

```diff
@@ -0,0 +1,391 @@
+
+        Returns:
+            Union[float, None]: 検出した直線と機体の中心線とのなす角、直線を検出できなかった場合は`None`を返す.
+        """
+        now = datetime.datetime.now()
+        self.__debug_time = now.strftime("%Y-%m-%d_%H-%M-%S.%f")
+
+        img = self.__camera_interface.capture_image()
+
+        if img is None:
+            return None
+
+        self.__save_debug_img(img, debug_img_fname_prefix, "captured.png")
+
+        # 射影変換
+        img_transformed = self.get_transformed_image(img)
+        self.__save_debug_img(img_transformed, debug_img_fname_prefix, "transformed.png")
+
+        # 画像から黒線を抽出(白黒画像に変換)
+        binary_img = BlackExtractor.extract_black(img_transformed)
+
+        # 線分を検出(線分上の2点座標[x1,y1,x2,y2])
+        detected_line = self.detect_line_segment(binary_img)
+
+        if detected_line is None:
+            return None
+
+        # 単位ベクトルに変換[x1-x2, y1-y2]
+        a = np.array([detected_line[0] - detected_line[2],
+                      detected_line[1] - detected_line[3]])
+        x = np.linalg.norm(a)
+        your_line = a / x
+
+        # 走行体のベクトル
+        y_axis = np.array([0, 1])
+
+        # 内積を求める
+        dot_product = np.dot(y_axis, your_line)
+
+        # なす角を求める
+        angle_2_y = np.arccos(dot_product) * 180 / np.pi
+
+        if angle_2_y > 90:
+            angle_2_y -= 180
+
+        if self.__debug:
+            # 検出画像の保存
+            debug_img = self.draw_detected_lines_on_the_image(
+                img_transformed, detected_line, angle_2_y)
+            self.__save_debug_img(debug_img, debug_img_fname_prefix, "detected.png")
+
+        return float(angle_2_y)
+
+    def get_transformed_image(self, img: np.ndarray, borderValue=(255, 255, 255)) -> np.ndarray:
+        """射影変換を行った画像を取得する.
+
+        Args:
+            img (np.ndarray): 射影変換前の画像.
+            borderValue (tuple, optional): 射影変換後の画像における余白部分の色. Defaults to (255, 255, 255).
+
+        Returns:
+            np.ndarray: 射影変換後の画像.
+        """
+        return cv2.warpPerspective(
+            img, self.__trans_mat, (img.shape[1], img.shape[0]), borderValue=borderValue)
+
+    def detect_line_segment(
+        self,
+        img,
+        length_threshold_mm: int = 145,
+        distance_threshold: float = 1.41421356,
+        canny_th1: int = 50,
+        canny_th2: int = 50,
+        canny_aperture_size: int = 7,  # 3~7
+        do_merge: bool = False
+    ) -> Union[np.ndarray, None]:
+        """画像上の線分を検出する.
+
+        Args:
+            img (np.ndarray): 射影変換後の画像.
+
+            NOTE:以下の引数の詳細: https://emotionexplorer.blog.fc2.com/blog-entry-128.html
+                                  https://nsr-9.hatenablog.jp/entry/2021/08/12/200000
+            length_threshold_mm (int): 線分を検出する際の長さ閾値(これより短い線は除外)
+                                       検出したい線分の長さ = 交点to中点-ブロックの半径
+                                                          = 175 - 30
+                                                          = 145
```

Comment:
交点の半径に内直径の値を使ってるのはなぜ？
あと、昨年は交点to中点は交点の半径も含んでいたけど、今年の応用動作に交点to中点に交点内移動は含んでないかな...？

Reply:
@kodama0720 @kawanoichi 
まず、確認したいこととして、交点って、サークルと直線の交わる部分なのか、サークルの中心なのかどっちなんでしょうか？
児玉さんと作業をしていたときは、サークルの中心が交点だとおもっていたのですが、
ブロックエリアのコマンドを作った木村君に聞いたところ、サークルと直線の交わる部分が交点ではないかと思ってます。
その前提で話を進めます。

確かに交点の半径に内直径の値を使ってるのは僕も疑問に思っていて、というか、検出したい線分の長さ = 交点to中点-ブロックの半径という式じゃなくて、検出したい線分の長さ = 交点to交点の線分の長さの半分で良いのではないかと思ってるのですが、どうなんでしょうか？

交点to中点に交点内移動は含んでいると思います。木村君に確認済みで、聞いたところ、まだmainブランチに更新していないInCrossLeft.cppなどがあるらしく、そのファイルには交点内移動の処理を追加しているそうです。

Reply:
CrossToMidの中でInCrossLeft(Right)してたね。すみません勘違いしていました。
```
検出したい線分の長さ = 交点to交点の線分の長さの半分（※交点サークル内の距離は除く）
```
```suggestion
                                       検出したい線分の長さ = 交点to交点の線分の長さの半分（※交点サークル内の距離は除く）
                                                          = 125？
```
みたいな感じのコメントにすれば分かりやすい？
「交点の半径に内直径の値を使ってるのは僕も疑問に思っている」ってのは @kodama0720 が担当してる場所ってことかな？

Reply:
> 
> まず、確認したいこととして、交点って、サークルと直線の交わる部分なのか、サークルの中心なのかどっちなんでしょうか？
> 

個人の認識になるけど一応。
※調整では違ったよってことならごめんなさい。


「交点」は、昨年(下図参照)も一昨年(交点to中点の参考元)も、線が交わってる円(塗りつぶしの差はあれど)の事のはず。交点サークルの略。

ただ昨年は、交点からの移動は、図にあるような「エッジの交差点(開発ではこれも略して交点って呼んでたかも)」から始めて、交点への移動は、エッジの交差点で終わるようにしてた。

一昨年は、交点からの移動は、交点とラインの接点から始めて、交点への移動は、交点とラインの接点で終わるようにしてた。
だから、交点to〇〇の交点は、サークルと交わる部分とか、サークルの中心とかではなく、サークルそのもの。

![image](https://github.com/KatLab-MiyazakiUniv/etrobocon2023/assets/83441177/4bac31d0-8be4-403a-b061-b30709f59eaa)

Reply:
児玉さんが担当していたというか、僕も一緒にやってたんですけど、何となくで分かってるつもりになってました。

Reply:
分からなかったら聞いてね